### PR TITLE
Upload image spring boot

### PR DESCRIPTION
**Pull request by veraivan**
## Description

The changes were made in the backend, so that the images are stored on the server (in this case in the local environment, once the system is on the Internet an external server is used). Then only the url of the image corresponding to the candidate is stored in the database. 

## Motivation and Context

The problem that we had before is that the binary code of the image was stored in the database and that at the moment of bringing a candidate in the JSON becomes very big being unnecessary and less effective that technique. Therefore, with this new method, we can store only the image url and we make the request to the server to return the image, thus being more efficient. 

## How Has This Been Tested?

The changes were tested, creating a candidate and adding a profile photo, once created the image is saved in an internal folder on the local machine called profile_images and then through the url is made the request to the backend to return the image. 

### Do you have any tests?
- [ ] Yes
- [x] No

### CODE REVIEW

File: back-end/src/main/java/com/swnat/controller/CandidateController.java [hunk 1]
```diff
@@ -46,5 +72,43 @@ public Candidate updateCandidate(@Valid @RequestBody Candidate candidate, @PathV
...
+        StringBuilder builder = new StringBuilder();
+        builder.append(directory);
+        builder.append(File.separator);
+        builder.append(image.getOriginalFilename());
+
+        byte [] filebytes = image.getBytes();
+        Path path = Paths.get(builder.toString());
+        Files.write(path, filebytes);
+
+        return "http://localhost:8080/api/v1/uploads/" + image.getOriginalFilename();
```
Comment (luifer-villalba) [inline back-end/src/main/java/com/swnat/controller/CandidateController.java]: FYI. Yesterday we had a meeting with @veraivan about this method and how it just going to work for a dev environment only. He's working on a solution.

**Review by luifer-villalba — CHANGES_REQUESTED** [review]

File: back-end/src/main/java/com/swnat/controller/CandidateController.java (hunk 1 above)
Comment (veraivan) [reply back-end/src/main/java/com/swnat/controller/CandidateController.java]: Solution completed :+1:. That line of code was modified and now the image is built in the front-end with a new method added.

Comment (veraivan) [reply back-end/src/main/java/com/swnat/controller/CandidateController.java]: > FYI. Yesterday we had a meeting with @veraivan about this method and how it just going to work for a dev environment only. He's working on a solution.

Solution completed :+1:. That line of code was modified and now the image is built in the front-end with a new method added.

File: front-end/src/app/candidate/data-candidate/data-candidate.component.ts [hunk 2]
```diff
@@ -117,7 +115,11 @@ export class DataCandidateComponent implements OnInit {
       return;
     } else {
       if (this.isNew) {
-        this.candidateService.addCandidate(this.dataCandidateForm.value, this.imageBinary).subscribe(data => {
+        this.formCandidate = new FormData();
+        this.formCandidate.append('candidate', JSON.stringify(this.dataCandidateForm.value));
+        this.formCandidate.append('imageProfile', this.fileUpload);
+        this.candidateService.addCandidate(this.formCandidate).subscribe(data => {
+          console.log(data);
```
Comment (luifer-villalba) [inline front-end/src/app/candidate/data-candidate/data-candidate.component.ts]: Delete console log

**Review by luifer-villalba — CHANGES_REQUESTED** [review]
The code tested and verified. Only a console log to delete and that's it for me.

File: front-end/src/app/candidate/data-candidate/data-candidate.component.ts (hunk 2 above)
Comment (veraivan) [reply front-end/src/app/candidate/data-candidate/data-candidate.component.ts]: The console log was removed :+1:

Comment (veraivan) [reply front-end/src/app/candidate/data-candidate/data-candidate.component.ts]: > Delete console log

> The console log was removed :+1:

Comment (veraivan) [reply front-end/src/app/candidate/data-candidate/data-candidate.component.ts]: > Delete console log
 
The console log was removed :+1:

**Review by luifer-villalba — DISMISSED** [review]

**Review by mauri-medina — COMMENTED** [review]
Great work! it works perfectly 👍 
I just have some suggestions that i think can make the code cleaner and more maintainable
-  Move the methods uploadImage and removeImage to CandidateService, right now they are in the controller and i don't think this is a controller job
- Create a global variable to store the new dir path and concatenate it with the user.dir. 
This change will help in the future if someones wants to change the route, it will be changed only in one place and not in all the methods that use it.

```
    private static final String fileUrl = File.separator + "src" +
            File.separator + "main" +
            File.separator + "java" +
            File.separator + "com" +
            File.separator + "swnat" +
            File.separator + "generic_files";

yourMethod( ){
   String directory = System.getProperty("user.dir") + fileUrl;
}
```

- Change the path of the new directory to be inside the resources directory of the project. 
I think that since this is an open source project its better not to enforce our package tree, someone could fork the project and change the pom groupId to `com.my-company` and an error will occur because the application will not find the com.swnat directory

Let me know if you have any questions or suggestions

**Comment by veraivan** [comment]
> Great work! it works perfectly 
> I just have some suggestions that i think can make the code cleaner and more maintainable
> 
> * Move the methods uploadImage and removeImage to CandidateService, right now they are in the controller and i don't think this is a controller job
> * Create a global variable to store the new dir path and concatenate it with the user.dir.
>   This change will help in the future if someones wants to change the route, it will be changed only in one place and not in all the methods that use it.
> 
> ```
>     private static final String fileUrl = File.separator + "src" +
>             File.separator + "main" +
>             File.separator + "java" +
>             File.separator + "com" +
>             File.separator + "swnat" +
>             File.separator + "generic_files";
> 
> yourMethod( ){
>    String directory = System.getProperty("user.dir") + fileUrl;
> }
> ```
> 
> * Change the path of the new directory to be inside the resources directory of the project.
>   I think that since this is an open source project its better not to enforce our package tree, someone could fork the project and change the pom groupId to `com.my-company` and an error will occur because the application will not find the com.swnat directory
> 
> Let me know if you have any questions or suggestions

Very good recommendations, thank you. The changes were made :+1:

**Review by mauri-medina — DISMISSED** [review]
It Works great! Thank you for the changes!

**Review by luifer-villalba — DISMISSED** [review]
Code tested and approved. Well done!

**Comment by luifer-villalba** [comment]
Well. I think we need to resolve conflicts to merge the branch @veraivan.

File: front-end/src/app/candidate/data-candidate/data-candidate.component.ts [hunk 3]
```diff
@@ -179,4 +183,14 @@ export class DataCandidateComponent implements OnInit {
       this.imageBinary = null;
```
Comment (mauri-medina) [inline front-end/src/app/candidate/data-candidate/data-candidate.component.ts:183]: I had to remove this line to be able to run the project

**Review by mauri-medina — COMMENTED** [review]
Thanks for the merge

File: front-end/src/app/candidate/data-candidate/data-candidate.component.ts (hunk 3 above)
Comment (veraivan) [reply front-end/src/app/candidate/data-candidate/data-candidate.component.ts:183]: > I had to remove this line to be able to run the project

No problem

**Review by luifer-villalba — APPROVED** [review]

**Review by mauri-medina — APPROVED** [review]

**Pull request merged**